internal/portforward/loop.go

```diff
@@ -1,64 +1,139 @@
-	logger Logger, puid, pgid int) *Loop {
-	start := make(chan struct{})
-	running := make(chan models.LoopStatus)
-	stop := make(chan struct{})
-	stopped := make(chan struct{})
-
-	statusManager := loopstate.New(constants.Stopped, start, running, stop, stopped)
-	state := state.New(statusManager, settings)
-
+	logger Logger, uid, gid int) *Loop {
 	return &Loop{
-		statusManager: statusManager,
-		state:         state,
-		puid:          puid,
-		pgid:          pgid,
-		// Objects
+		settings: service.Settings{
+			Settings: settings,
+		},
 		client:      client,
 		portAllower: portAllower,
 		logger:      logger,
-		start:       start,
-		running:     running,
-		stop:        stop,
-		stopped:     stopped,
-		userTrigger: true,
-		backoffTime: defaultBackoffTime,
+		uid:         uid,
+		gid:         gid,
+	}
+}
+
+func (l *Loop) Start(_ context.Context) (runError <-chan error, _ error) {
+	l.runCtx, l.runCancel = context.WithCancel(context.Background())
+	runDone := make(chan struct{})
+	l.runDone = runDone
+
+	updatedSignal := make(chan struct{})
+	l.updatedSignal = updatedSignal
+	runErrorCh := make(chan error)
+
+	go l.run(l.runCtx, runDone, runErrorCh, updatedSignal)
+
+	return runErrorCh, nil
+}
+
+func (l *Loop) run(runCtx context.Context, runDone chan<- struct{},
+	runErrorCh chan<- error, updatedSignal <-chan struct{}) {
+	defer close(runDone)
+
+	var serviceRunError <-chan error
+	for {
+		select {
+		case <-runCtx.Done():
+			// Stop call takes care of stopping the service
+			return
+		case <-updatedSignal: // first and subsequent start trigger
+		case err := <-serviceRunError:
+			runErrorCh <- err
+			return
+		}
+
+		firstRun := l.service == nil
+		if !firstRun {
+			err := l.service.Stop()
+			if err != nil {
+				runErrorCh <- fmt.Errorf("stopping previous service: %w", err)
+				return
+			}
+		}
+
+		l.settingsMutex.RLock()
+		l.service = service.New(l.settings, l.client,
+			l.portAllower, l.logger, l.uid, l.gid)
+		l.settingsMutex.RUnlock()
+
+		var err error
+		serviceRunError, err = l.service.Start(runCtx)
+		if err != nil {
+			if runCtx.Err() == nil { // crashed but NOT stopped
+				runErrorCh <- fmt.Errorf("starting new service: %w", err)
+			}
+			return
+		}
+	}
+}
+
+func (l *Loop) UpdateWith(partialUpdate service.Settings) (err error) {
+	l.settingsMutex.Lock()
+	err = l.settings.UpdateWith(partialUpdate)
+	l.settingsMutex.Unlock()
+	if err != nil {
+		return err
+	}
+
+	select {
+	case l.updatedSignal <- struct{}{}:
+		// Settings are validated and if the service fails to start
+		// or crashes at runtime, the loop will stop and signal its
+		// parent goroutine. Settings validation should be the only
+		// error feedback for the caller of `Update`.
+		return nil
+	case <-l.runCtx.Done():
+		// loop has been stopped, no update can be done
+		return l.runCtx.Err()
+	}
+}
+
+func (l *Loop) Stop() (err error) {
+	l.runCancel()
+	<-l.runDone
+
+	if l.service != nil {
+		return l.service.Stop()
+	}
```

Comment:
this should be moved to execute in the `run` function, in case Stop is called when Start is running in the `run` function, which would cause a Stop on an already start-aborted service.

Reply:
Keeping this as is, so we access the service stop error;
calling Stop on an start-aborted service will simply return `nil` which is fine in this case.